xarray/namedarray/core.py

```diff
@@ -0,0 +1,443 @@
+        --------
+        numpy.ndarray.imag
+        """
+        return self._replace(data=self.data.imag)
+
+    def __dask_tokenize__(self: T_NamedArray):
+        # Use v.data, instead of v._data, in order to cope with the wrappers
+        # around NetCDF and the like
+        from dask.base import normalize_token
+
+        return normalize_token((type(self), self._dims, self.data, self.attrs))
+
+    def __dask_graph__(self: T_NamedArray):
+        return self._data.__dask_graph__() if is_duck_dask_array(self._data) else None
+
+    def __dask_keys__(self: T_NamedArray):
+        return self._data.__dask_keys__()
+
+    def __dask_layers__(self: T_NamedArray):
+        return self._data.__dask_layers__()
+
+    @property
+    def __dask_optimize__(self: T_NamedArray) -> typing.Callable:
+        return self._data.__dask_optimize__
+
+    @property
+    def __dask_scheduler__(self: T_NamedArray) -> typing.Callable:
+        return self._data.__dask_scheduler__
+
+    def __dask_postcompute__(
+        self: T_NamedArray,
+    ) -> tuple[typing.Callable, tuple[typing.Any, ...]]:
+        array_func, array_args = self._data.__dask_postcompute__()
+        return self._dask_finalize, (array_func,) + array_args
+
+    def __dask_postpersist__(
+        self: T_NamedArray,
+    ) -> tuple[typing.Callable, tuple[typing.Any, ...]]:
+        array_func, array_args = self._data.__dask_postpersist__()
+        return self._dask_finalize, (array_func,) + array_args
+
+    def _dask_finalize(
+        self: T_NamedArray, results, array_func, *args, **kwargs
+    ) -> T_NamedArray:
+        data = array_func(results, *args, **kwargs)
+        return type(self)(self._dims, data, attrs=self._attrs)
+
+    @property
+    def chunks(self: T_NamedArray) -> tuple[tuple[int, ...], ...] | None:
+        """
+        Tuple of block lengths for this dataarray's data, in order of dimensions, or None if
+        the underlying data is not a dask array.
+
+        See Also
+        --------
+        Variable.chunk
+        Variable.chunksizes
+        xarray.unify_chunks
+        """
+        return getattr(self._data, "chunks", None)
+
+    @property
+    def chunksizes(
+        self: T_NamedArray,
+    ) -> typing.Mapping[typing.Any, tuple[int, ...]]:
+        """
+        Mapping from dimension names to block lengths for this variable's data, or None if
+        the underlying data is not a dask array.
+        Cannot be modified directly, but can be modified by calling .chunk().
+
+        Differs from variable.chunks because it returns a mapping of dimensions to chunk shapes
+        instead of a tuple of chunk shapes.
+
+        See Also
+        --------
+        Variable.chunk
+        Variable.chunks
+        xarray.unify_chunks
+        """
+        if hasattr(self._data, "chunks"):
+            return dict(zip(self.dims, self.data.chunks))
+        else:
+            return {}
+
+    @property
+    def sizes(self: typing.Any) -> dict[Hashable, int]:
```

Comment:
i added a method with a slightly different return type `dict[Hashable, int]` compared to the original `Frozen[Hashable, int]`. unfortunately, this change breaks the typing.


```
xarray/core/variable.py:314: error: Definition of "sizes" in base class "NamedArray" is incompatible with definition in base class "AbstractArray"  [misc]
```

@Illviljan, i would appreciate your assistance once again ;)

Reply:
How about just removing the Frozen from abstractarray? There's no point there either.

https://github.com/pydata/xarray/blob/8c21376aa94deb91cb10e78f56ecd68aa7fe90fa/xarray/core/common.py#L225

Reply:
i tried, but it ended up creating more problems

```bash
xarray/core/common.py: note: In member "where" of class "DataWithCoords":
xarray/core/common.py:1161: error: Value of type variable "T_Alignable" of "align" cannot be "T_DataWithCoords | Dataset | DataArray"  [type-var]
xarray/core/common.py:1161: note: Error code "type-var" not covered by "type: ignore" comment
xarray/core/groupby.py: note: In function "_resolve_group":
xarray/core/groupby.py:612: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/core/groupby.py: note: In member "sizes" of class "GroupBy":
xarray/core/groupby.py:760: error: Incompatible types in assignment (expression has type "dict[Hashable, int]", variable has type "Frozen[Hashable, int] | None")  [assignment]
xarray/core/groupby.py:764: error: Incompatible return value type (got "Frozen[Hashable, int] | None", expected "Frozen[Hashable, int]")  [return-value]
xarray/core/dataset.py: note: In member "broadcast_like" of class "Dataset":
xarray/core/dataset.py:3330: error: Value of type variable "T_Alignable" of "align" cannot be "Dataset | DataArray"  [type-var]
xarray/core/dataset.py: note: In member "_binary_op" of class "Dataset":
xarray/core/dataset.py:7496: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray | Dataset"  [type-var]
xarray/core/dataset.py:7496: note: Error code "type-var" not covered by "type: ignore" comment
xarray/core/dataset.py: note: In member "curvefit" of class "Dataset":
xarray/core/dataset.py:9662: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/core/dataarray.py: note: In member "broadcast_like" of class "DataArray":
xarray/core/dataarray.py:1802: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray | Dataset"  [type-var]
xarray/core/dataarray.py:1806: error: Value of type variable "T_Alignable" of "_broadcast_helper" cannot be "T_DataArray"  [type-var]
xarray/core/dataarray.py: note: In member "reindex_like" of class "DataArray":
xarray/core/dataarray.py:2006: error: Value of type variable "T_Alignable" of "reindex_like" cannot be "T_DataArray"  [type-var]
xarray/core/dataarray.py: note: In member "reindex" of class "DataArray":
xarray/core/dataarray.py:2097: error: Value of type variable "T_Alignable" of "reindex" cannot be "T_DataArray"  [type-var]
xarray/core/computation.py: note: In function "_cov_corr":
xarray/core/computation.py:1437: error: Value of type variable "T_Alignable" of "align" cannot be "T_DataArray"  [type-var]
xarray/core/computation.py: note: In function "cross":
xarray/core/computation.py:1635: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/plot/dataset_plot.py: note: In function "quiver":
xarray/plot/dataset_plot.py:456: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/plot/dataset_plot.py: note: In function "streamplot":
xarray/plot/dataset_plot.py:626: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/plot/dataarray_plot.py: note: In function "_prepare_plot1d_data":
xarray/plot/dataarray_plot.py:212: error: Value of type variable "T_Alignable" of "broadcast" cannot be "T_DataArray"  [type-var]
xarray/core/weighted.py: note: In member "_weighted_quantile" of class "Weighted":
xarray/core/weighted.py:428: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/core/weighted.py:429: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_array_api.py: note: In function "test_broadcast":
xarray/tests/test_array_api.py:72: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_array_api.py:73: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:278: error: Unused "type: ignore" comment  [unused-ignore]
xarray/tests/test_dataarray.py: note: In member "test_broadcast_like" of class "TestDataArray":
xarray/tests/test_dataarray.py:1682: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:1691: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align" of class "TestDataArray":
xarray/tests/test_dataarray.py:2987: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_dtype" of class "TestDataArray":
xarray/tests/test_dataarray.py:2997: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_copy" of class "TestDataArray":
xarray/tests/test_dataarray.py:3007: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3012: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3019: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3023: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_override" of class "TestDataArray":
xarray/tests/test_dataarray.py:3041: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3045: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3049: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3058: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_exclude" of class "TestDataArray":
xarray/tests/test_dataarray.py:3086: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_indexes" of class "TestDataArray":
xarray/tests/test_dataarray.py:3106: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3112: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_without_indexes_exclude" of class "TestDataArray":
xarray/tests/test_dataarray.py:3118: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_mixed_indexes" of class "TestDataArray":
xarray/tests/test_dataarray.py:3125: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3129: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_without_indexes_errors" of class "TestDataArray":
xarray/tests/test_dataarray.py:3138: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3144: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_align_str_dtype" of class "TestDataArray":
xarray/tests/test_dataarray.py:3160: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_broadcast_arrays" of class "TestDataArray":
xarray/tests/test_dataarray.py:3171: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3180: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_broadcast_arrays_misaligned" of class "TestDataArray":
xarray/tests/test_dataarray.py:3198: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_broadcast_arrays_nocopy" of class "TestDataArray":
xarray/tests/test_dataarray.py:3210: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:3216: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_broadcast_arrays_exclude" of class "TestDataArray":
xarray/tests/test_dataarray.py:3225: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_dot_align_coords" of class "TestDataArray":
xarray/tests/test_dataarray.py:4037: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py:4049: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_dataarray.py: note: In member "test_binary_op_join_setting" of class "TestDataArray":
xarray/tests/test_dataarray.py:4113: error: Value of type variable "T_Alignable" of "align" cannot be "DataArray"  [type-var]
xarray/tests/test_interp.py: note: In function "test_decompose":
xarray/tests/test_interp.py:765: error: Value of type variable "T_Alignable" of "broadcast" cannot be "DataArray"  [type-var]
```

Reply:
Aha now I get it!

https://github.com/pydata/xarray/blob/3ace2fb4612d4bc1cbce6fa22fe3954a0e06599e/xarray/core/types.py#L109-L111

Need to change the type hints there as well.

Reply:
wow! 🎉 good catch.  thank you for looking into this